p9cpu/crates/libp9cpu/src/server.rs

```diff
@@ -1,63 +1,314 @@
+    #[error("Cannot open pty device")]
+    OpenPtyFail(#[from] nix::Error),
+    #[error("Cannot clone file descriptor: {0}")]
+    FdCloneFail(#[source] std::io::Error),
+    #[error("Cannot create directory: {0}")]
+    MkDir(#[source] std::io::Error),
+    #[error("Invalid FsTab: {0}")]
+    InvalidFsTab(String),
+    #[error("Cannot bind listener: {0}")]
+    BindFail(#[source] std::io::Error),
+    #[error("9p forward not setup")]
+    No9pPort,
+    #[error("String contains null: {0:?}")]
+    StringContainsNull(#[from] std::ffi::NulError),
+    #[error("Channel closed")]
+    ChannelClosed,
+}
+
+impl<T> From<mpsc::error::SendError<T>> for Error {
+    fn from(_: mpsc::error::SendError<T>) -> Self {
+        Self::ChannelClosed
+    }
+}
 
 #[derive(Debug)]
-pub struct Server<I> {
-    _phantom: Option<I>,
+pub struct Server<I, L> {
+    launcher: L,
+    sessions: Arc<RwLock<HashMap<I, Session>>>,
+    pending: Arc<RwLock<HashMap<I, PendingSession>>>,
 }
 
-impl<I> Default for Server<I> {
-    fn default() -> Self {
-        Self { _phantom: None }
+impl<I, L> Server<I, L> {
+    pub fn new(launcher: L) -> Self {
+        Self {
+            sessions: Arc::new(RwLock::new(HashMap::new())),
+            pending: Arc::new(RwLock::new(HashMap::new())),
+            launcher,
+        }
     }
 }
 
-impl<I> Server<I> {
-    pub async fn start(&self, _command: cmd::Cmd, _sid: I) -> Result<(), Error> {
-        unimplemented!()
+impl<I, L> Server<I, L>
+where
+    L: crate::launcher::Launch,
+    I: Eq + Hash + Debug + Display + Sync + Send + Clone + 'static,
+{
+    async fn get_session<O, R>(&self, sid: &I, op: O) -> Result<R, Error>
+    where
+        O: Fn(&Session) -> R,
+    {
+        let sessions = self.sessions.read().await;
+        let info = sessions.get(sid).ok_or(Error::SessionNotExist)?;
+        Ok(op(info))
+    }
+
+    async fn copy_to(
+        src: &mut (impl AsyncRead + Unpin),
+        tx: mpsc::Sender<Vec<u8>>,
+    ) -> Result<(), Error> {
+        loop {
+            let mut buf = vec![0; 128];
+            let len = src.read(&mut buf).await.map_err(Error::IoErr)?;
+            if len == 0 {
+                break;
+            }
+            buf.truncate(len);
+            tx.send(buf).await?;
+        }
+        Ok(())
+    }
+
+    pub async fn start(&self, cmd: cmd::Cmd, sid: I) -> Result<(), Error> {
+        let Some(PendingSession { ninep }) = self.pending.write().await.remove(&sid) else {
+            return Err(Error::SessionNotExist);
+        };
+        let mut sessions = self.sessions.write().await;
+        if sessions.contains_key(&sid) {
+            return Err(Error::DuplicateId);
+        }
+        let mut handles = vec![];
+        let mut ninep_port = None;
+        if let Some((port, handle)) = ninep.write().await.take() {
+            ninep_port = Some(port);
+            handles.push(handle);
+        }
+
+        let uds_path = format!("/tmp/p9cpud-launcher-{}", &sid);
```

Comment:
is this in a private namespace?

Reply:
In the new commit, I switched to a tcp port to avoid conflicts.